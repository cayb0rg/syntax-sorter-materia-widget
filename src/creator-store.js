
import React, { useReducer } from 'react'

let legendIdIncrement = 1

const init = {
	requireInit: true,
	currentIndex: 0,
	showTutorial: true,
	showHintModal: false,
	showFakeoutModal: false,
	showBankModal: false,
	showErrorModal: false,
	errorMsg: "",
	selectedTokenIndex: -1,
	selectedFakeoutIndex: -1,
	title: 'New Foreign Language Wiget',
	items: [{
		question: '',
		phrase: [],
		displayPref: 'word',
		checkPref: 'no',
		numChecks: 1,
		hint: '',
		fakeoutPref: 'no',
		fakeout: []
	}],
	legend: [
		{
			id: legendIdIncrement,
			color: '#FF0000',
			name: 'Part of Speech'
		}
	],
	numAsk: 1,
	askLimit: "no",
	showLegend: false,
	legendColorPickerTarget: -1
}
const store = React.createContext(init)
const { Provider } = store

// not a reducer, just takes qset data and transforms it into store format
const importFromQset = (qset) => {
	let items = qset.items.map((item) => {
		return {
			question: item.questions[0].text,
			phrase: item.answers[0].options.phrase,
			displayPref: item.options.displayPref,
			checkPref: item.options.checkPref,
			numChecks: item.options.numChecks,
			hint: item.options.hint,
			fakeoutPref: item.options.fakeoutPref,
			fakeout: item.options.fakeout
		}
	})

	legendIdIncrement = qset.options.legend.length

	return {
		items: items,
		legend: qset.options.legend,
		numAsk: qset.options.numAsk,
		askLimit: qset.options.askLimit
	}
}

const questionItemReducer = (items, action) => {
	switch (action.type) {
		case 'add_new_question':
			return [...items, {
				question: '',
				phrase: [],
				displayPref: 'word',
				checkPref: 'no',
				numChecks: 0,
				hint: '',
				fakeoutPref: 'no',
				fakeout: []
			}]
		case 'update_question_text':
			return items.map((item, index) => {
				if (index == action.payload.index) {
					return {
						...item,
						question: action.payload.text
					}
				}
				else return item
			})
		case 'remove_question':
			return [
				...items.slice(0, action.payload),
				...items.slice(action.payload+1)
			]
		case 'phrase_token_to_input':
		case 'phrase_input_to_token':
		case 'phrase_token_type_select':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {
						...item,
						phrase: phraseReducer(item.phrase, action)
					}
				}
				else return item
			})
		case 'fakeout_token_to_input':
		case 'fakeout_input_to_token':
		case 'fakeout_token_type_select':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {
						...item,
						fakeout: fakeoutReducer(item.fakeout, action)
					}
				}
				else return item
			})
		case 'update_display_pref':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {
						...item,
						displayPref: action.payload.pref
					}
				}
				else return item
			})
		case 'update_check_pref':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {
						...item,
						checkPref: action.payload.pref
					}
				}
				else return item
			})
		case 'update_num_checks':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {
						...item,
						numChecks: parseInt(action.payload.pref)
					}
				}
				else return item
			})
		case 'update_hint':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {
						...item,
						hint: action.payload.pref
					}
				}
				else return item
			})
		case 'update_fakeout_pref':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {
						...item,
						fakeoutPref: action.payload.pref
					}
				}
				else return item
			})
		case 'remove_legend_item':
			return items.map((item) => {
				return {
					...item,
					phrase: phraseReducer(item.phrase, action)
				}
			})
		case 'fakeout_remove_legend_item':
			return items.map((item) => {
				return {
					...item,
					fakeout: fakeoutReducer(item.fakeout, action)
				}
			})
		default:
			throw new Error('Question item reducer: this action type was not defined')
	}
}

const phraseReducer = (phrase, action) => {
	switch (action.type) {
		case 'phrase_token_to_input':
			return [
				...phrase.slice(0,action.payload.phraseIndex),
				...phrase.slice(action.payload.phraseIndex + 1)
			]
		case 'phrase_input_to_token':
			return [
				...phrase,
				{
					value: action.payload.text,
					legend: null
				}
			]
		case 'phrase_token_type_select':
			return phrase.map((token, index) => {
				if (index == action.payload.phraseIndex) {
					return {
						...token,
						legend: action.payload.selection
					}
				}
				else return token
			})
		case 'remove_legend_item':
			return phrase.map((token) => {
				if (token.legend == action.payload.id) {
					return {
						...token,
						legend: null
					}
				}
				else return token
			})
		default:
			throw new Error('Phrase item reducer: this action type was not defined')
	}
}

const fakeoutReducer = (fakeout, action) => {
	switch (action.type) {
		case 'fakeout_token_to_input':
			return [
				...fakeout.slice(0,action.payload.fakeoutIndex),
				...fakeout.slice(action.payload.fakeoutIndex + 1)
			]
		case 'fakeout_input_to_token':
			return [
				...fakeout,
				{
					value: action.payload.text,
					legend: null
				}
			]
		case 'fakeout_token_type_select':
			return fakeout.map((token, index) => {
				if (index == action.payload.fakeoutIndex) {
					return {
						...token,
						legend: action.payload.selection
					}
				}
				else return token
			})
		case 'fakeout_remove_legend_item':
			return fakeout.map((token) => {
				if (token.legend == action.payload.id) {
					return {
						...token,
						legend: null
					}
				}
				else return token
			})
		default:
			throw new Error('Phrase item reducer: this action type was not defined')
	}
}

const legendReducer = (legend, action) => {
	switch (action.type) {
		case 'add_legend_item':
			return [...legend, {
				id: ++legendIdIncrement,
				color: action.payload.color,
				name: action.payload.text
			}]
		case 'update_legend_item':
			return legend.map((term, index) => {
				if (index == action.payload.index) {
					return {
						...term,
						name: action.payload.text
					}
				}
				else return term
			})
		case 'remove_legend_item':
			return [
				...legend.slice(0, action.payload.index),
				...legend.slice(action.payload.index + 1)
			]
		case 'legend_color_picker_change':
			console.log("color picker change: index: " + action.payload.index + " and color: " + action.payload.color)
			return legend.map((term, index) => {
				if (index == action.payload.index) {
					return {
						...term,
						color: action.payload.color
					}
				}
				else return term
			})
		default:
			throw new Error('Legend reducer: this action type was not defined')
	}
}

const StateProvider = ( { children } ) => {
	const [state, dispatch] = useReducer((state, action) => {
		switch (action.type) {
			case 'init-new':
				return {...state, requireInit: false}
			case 'init-existing':
				let imported = importFromQset(action.payload.qset)
				return {...state, title: action.payload.title, items: imported.items, legend: imported.legend, numAsk: imported.numAsk, askLimit: imported.askLimit, requireInit: false}
			case 'dismiss_tutorial':
				return {...state, showTutorial: false}
			case 'update_title':
				return {...state, title: action.payload}
			case 'remove_question':
				if (state.items.length < 2) return state
				else return {...state, items: questionItemReducer(state.items, action), currentIndex: state.currentIndex - 1}
			case 'add_new_question':
			case 'update_question_text':
			case 'update_display_pref':
			case 'update_check_pref':
			case 'update_num_checks':
			case 'update_fakeout_pref':
			case 'update_fakeout':
			case 'update_hint':
			case 'phrase_token_to_input':
			case 'phrase_input_to_token':
				return {...state, items: questionItemReducer(state.items, action)}
			case 'fakeout_token_to_input':
			case 'fakeout_input_to_token':
				return {...state, items: questionItemReducer(state.items, action)}
			case 'phrase_token_type_select':
				return {...state, items: questionItemReducer(state.items, action), selectedTokenIndex: -1}
			case 'fakeout_token_type_select':
				return {...state, items: questionItemReducer(state.items, action), selectedFakeoutIndex: -1}
			case 'toggle_token_select':
				return {...state, selectedTokenIndex: action.payload != state.selectedTokenIndex ? action.payload : -1}
			case 'toggle_fakeout_select':
				return {...state, selectedFakeoutIndex: action.payload != state.selectedFakeoutIndex ? action.payload : -1}
			case 'select_question':
				return {...state, currentIndex: action.payload}
			case 'toggle_legend':
				return {...state, showLegend: !state.showLegend}
			case 'add_legend_item':
				return {...state, legend: legendReducer(state.legend, action)}
			case 'update_legend_item':
				return {...state, legend: legendReducer(state.legend, action)}
			case 'remove_legend_item':
				return {...state, items: questionItemReducer(state.items, action), legend: legendReducer(state.legend, action)}
			case 'fakeout_remove_legend_item':
				return {...state, items: questionItemReducer(state.items, action)}
			case 'legend_color_picker_toggle':
				if (state.legendColorPickerTarget != action.payload.index) return {...state, legendColorPickerTarget: action.payload.index}
				else return {...state, legendColorPickerTarget: -1}
			case 'legend_color_picker_change':
				return {...state, legend: legendReducer(state.legend, action), legendColorPickerTarget: -1}
			case 'toggle_hint_modal':
				return {...state, showHintModal: !state.showHintModal}
			case 'toggle_fakeout_modal':
				return {...state, showFakeoutModal: !state.showFakeoutModal}
			case 'toggle_bank_modal':
				return {...state, showBankModal: !state.showBankModal}
			case 'toggle_error_modal':
				return {...state, errorMsg: action.payload.error, showErrorModal: !state.showErrorModal}
			case 'update_num_ask':
				return {...state, numAsk: action.payload}
			case 'update_ask_limit':
				return {...state, askLimit: action.payload}
			default:
			  throw new Error('Base reducer: this action type was not defined')
		  }
	}, init)

	return <Provider value={{state, dispatch}}>{children}</Provider>
}

export {store, StateProvider }
