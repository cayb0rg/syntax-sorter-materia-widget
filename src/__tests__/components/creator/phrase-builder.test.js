import React from 'react'
import renderer from 'react-test-renderer'
import PhraseBuilder from '../../../components/creator/phrase-builder'
import Token from '../../../components/creator/token'
import { StateProvider } from '../../../creator-store'
import { act } from '@testing-library/react-hooks'

describe('Phrase Builder', () => {
	
	test('Phrase Builder component', () => {
		const component = renderer.create(
			<StateProvider>
				<PhraseBuilder />
			</StateProvider>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('It inputs a token', () => {
		const component = renderer.create(
			<StateProvider>
				<PhraseBuilder />
			</StateProvider>
		)

		const providerEl = component.root
		const tokenInput = providerEl.findByProps({ className: 'token-input'})

		act(() => {
			tokenInput.props.onKeyDown({which: 13, target: { value: 'antidisestablishmentarianism'}})
		})

		const tokens = providerEl.findAllByType(Token)
		expect(tokens[0].props.value).toBe('antidisestablishmentarianism')
	})

	test('It selects a legend type for the token', () => {
		const component = renderer.create(
			<StateProvider>
				<PhraseBuilder />
			</StateProvider>
		)

		const providerEl = component.root
		const tokenInput = providerEl.findByProps({ className: 'token-input' })
			
		act(() => {
			tokenInput.props.onKeyDown({which: 13, target: { value: 'antidisestablishmentarianism'}})
		})

		const tokens = providerEl.findAllByType(Token)

		expect(tokens[0].props.type).toBeNull()

		let tokenSelect = tokens[0].findByType('span')

		act(() => {
			tokenSelect.props.onClick()
		})

		expect(tokenSelect.props.className).toBe('token unassigned selected')

		let inputs = providerEl.findAllByType('input')
		inputs.forEach((input) => {
			if (input.props.name && input.props.value == '1') {
				act(() => {
					input.props.onChange({target: { value: '1' }})
				})
			}
		})

		expect(tokens[0].props.value).toBe('antidisestablishmentarianism')
		expect(tokens[0].props.type).toBe(1)
	})

	test('It converts a token back into input', () => {
		const component = renderer.create(
			<StateProvider>
				<PhraseBuilder />
			</StateProvider>
		)

		const providerEl = component.root
		const phraseBuilderEl = providerEl.findByType(PhraseBuilder)

		let inputs = phraseBuilderEl.findAllByType('input')

		inputs.forEach((input) => {

			if (input.props.className && input.props.className == 'token-input'){
				act(() => {
					input.props.onKeyDown({which: 13, target: { value: 'antidisestablishmentarianism'}})
				})
			}
		})

		let tokens = phraseBuilderEl.findAllByType(Token)
		expect(tokens.length).toBe(1)

		inputs.forEach((input) => {

			if (input.props.className && input.props.className == 'token-input'){
				act(() => {
					input.props.onKeyDown({which: 8, target: { value: ''}, preventDefault: jest.fn()})
				})

				tokens = phraseBuilderEl.findAllByType(Token)
				expect(tokens.length).toBe(0)
			}
		})
	})
})