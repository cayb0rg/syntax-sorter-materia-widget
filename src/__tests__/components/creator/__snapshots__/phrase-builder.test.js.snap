// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phrase Builder Phrase Builder component 1`] = `
<section
  className="card phrase-builder"
>
  <header>
    Phrase to Complete
  </header>
  <div
    className="token-tutorial show"
    onClick={[Function]}
  >
    <p>
      <span
        className="icon-notification"
      />
      Use the input to the left to create the individual 
      <span
        className="strong"
      >
        tokens
      </span>
       that will make up your phrase. Tokens can be a word, multiple words, a part of speech, grammar symbol, or any combination. The tokens will be randomly ordered when a student plays the widget.
    </p>
  </div>
  <div
    className="token-container"
  >
    <div
      className="token-input-container"
    >
      <input
        className="token-input"
        maxLength={200}
        onKeyDown={[Function]}
        placeholder="..."
      />
    </div>
  </div>
  <div
    className="token-type-selector "
  >
    <header>
      Select the corresponding legend item for this token:
    </header>
    <form
      id="tokenTypeSelection"
    >
      <label
        className=""
      >
        <input
          checked={false}
          name="token-type-selection"
          onChange={[Function]}
          type="radio"
          value={1}
        />
        <span
          className="color-radio"
          style={
            Object {
              "background": "#FF0000",
            }
          }
        />
        Part of Speech
      </label>
      <span
        className="legend-reminder"
      >
        Open the 
        <a
          href="#"
          onClick={[Function]}
        >
          Legend
        </a>
         to create additional token categories.
      </span>
    </form>
  </div>
</section>
`;
